### DISCUSSION TITLE
Chained withers / with-expression possible

### DISCUSSION THREAD

**Discussion by msedi** [dotnet/csharplang#6385] [Q&A] [answered]
I have a read-only structure where properties are dependent from other properties in the same structure. It would be completely sufficient to have everything set in the initialization using a compound initialization with `init` it is not possible for several reasons, because in the example order might play a role, so I cannot access the Blocksize within the initialization:

```cs
 var gridSize1 = new CudaGridBlockSize
            {
                BlocksizeX = 64,
                BlocksizeY = 1,
                BlocksizeZ = 1,
                GridsizeX = ((width - 1) / BlocksizeX / 4) + 1,
                GridsizeY = ((height - 1) / BlocksizeY) + 1,
                GridsizeZ = ((depth - 1) / BlocksizeZ) + 1
            } 
```

so I came up with the idea of using withers but found that also here I have no access to the original. Is this even possible?

```cs
 var gridSize1 = new CudaGridBlockSize
            {
                BlocksizeX = 64,
                BlocksizeY = 1,
                BlocksizeZ = 1
            } 
            with
            {
                // Accessing gridsize1 does not work
                GridsizeX = ((width - 1) / gridSize1.BlocksizeX / 4) + 1,
                GridsizeY = ((height - 1) / gridSize1.BlocksizeY) + 1,
                GridsizeZ = ((depth - 1) / gridSize1.BlocksizeZ) + 1
            };
```

**Comment by HaloFour** [comment, marked as answer]
You can't access `gridSize1` until after it has been assigned, which in your expression isn't until after the `with` expression which is why it's not accessible there.

If you assign the variable first you can use `with` in this manner:

```cs
 var gridSize1 = new CudaGridBlockSize
            {
                BlocksizeX = 64,
                BlocksizeY = 1,
                BlocksizeZ = 1
            } ;
  gridSize1 = gridSize1 with
            {
                GridsizeX = ((width - 1) / gridSize1.BlocksizeX / 4) + 1,
                GridsizeY = ((height - 1) / gridSize1.BlocksizeY) + 1,
                GridsizeZ = ((depth - 1) / gridSize1.BlocksizeZ) + 1
            };
```

There is a separate issue where improvements to initializers/withers are being considered but I don't think it includes attempting to reference other properties like this:  #5176

**Reply by msedi** [reply]
@HaloFour. Thanks for the answer. I thought so and thats how I do it now. I found this problem when refactoring code to use readonly-structs. Do you think I should attach some proposal to #5176.
